### PR TITLE
ports/stm32/flash.c: WB55 flash locking.

### PR DESCRIPTION
**Pull request by jimmo**
(Extends from #6594, just the last three commits are relevant to this PR)

The WB55 wireless core shares the same internal flash as the app core, and so when the app core does flash erase or write operations, then the wireless core will stall as the bus is locked.

This adds a BLE multitest to demonstrate the issue, and then fixes it by:
 - Processing HCI data as quickly as possible in the RX IRQ (this requires that we no longer support btstack on WB55 for now)
 - Implementing the flash locking algorithm described in AN5289 Rev 3, Figure 10.

### CODE REVIEW

File: tests/multi_bluetooth/stress_log_filesystem.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
+# Test BLE GAP connect/disconnect
```
Comment (dpgeorge) [inline tests/multi_bluetooth/stress_log_filesystem.py:1]: this comment needs to be changed (cut and paste error?)

File: tests/multi_bluetooth/stress_log_filesystem.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    elif event == _IRQ_PERIPHERAL_CONNECT:
+        print("_IRQ_PERIPHERAL_CONNECT")
+        waiting_events[event] = data[0]
+    elif event == _IRQ_PERIPHERAL_DISCONNECT:
+        print("_IRQ_PERIPHERAL_DISCONNECT")
+    elif event == _IRQ_GATTC_SERVICE_RESULT:
+        # conn_handle, start_handle, end_handle, uuid = data
+        if data[-1] == SERVICE_UUID:
+            print("_IRQ_GATTC_SERVICE_RESULT", data[3])
+            waiting_events[event] = (data[1], data[2],)
```
Comment (dpgeorge) [inline tests/multi_bluetooth/stress_log_filesystem.py:73]: please remove trailing comma on RHS, and then run `black -l99` on this file (to fix other formatting issues)

File: tests/multi_bluetooth/stress_log_filesystem.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        waiting_events[event] = None
+
+
+def wait_for_event(event, timeout_ms):
+    t0 = time.ticks_ms()
+    while time.ticks_diff(time.ticks_ms(), t0) < timeout_ms:
+        periodic_log_write()
+        if event in waiting_events:
+            result = waiting_events[event]
+            del waiting_events[event]
```
Comment (dpgeorge) [inline tests/multi_bluetooth/stress_log_filesystem.py:108]: I think `waiting_events.pop(event)` is a better way to write this, it's simpler and more atomic (especially if IRQs can change this dict).

Could even write this as:
```python
if result := waiting_events.pop(event, -1) != -1:
    return result
```
But not sure if it's good to put a dependency on the walrus operator....

File: tests/multi_bluetooth/stress_log_filesystem.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            ble.gap_advertise(50_000, b"\x02\x01\x06\x04\xffMPY")
+            # Wait for central to connect, do a sequence of read/write, then disconnect.
+            wait_for_event(_IRQ_CENTRAL_CONNECT, TIMEOUT_MS)
+            for op in range(4):
+                wait_for_event(_IRQ_GATTS_READ_REQUEST, TIMEOUT_MS)
+                wait_for_event(_IRQ_GATTS_WRITE, TIMEOUT_MS)
+            wait_for_event(_IRQ_CENTRAL_DISCONNECT, 2*TIMEOUT_MS)
+    finally:
+        ble.active(0)
+        log_file.close()
```
Comment (dpgeorge) [inline tests/multi_bluetooth/stress_log_filesystem.py:136]: would it make sense for the test to re-open the file and print out its contents at the end here, for further verification it ran correctly?

File: ports/stm32/rfcore.c [hunk 5]
```diff
@@ -597,7 +600,8 @@ void rfcore_ble_hci_cmd(size_t len, const uint8_t *src) {
 
 void rfcore_ble_check_msg(int (*cb)(void *, const uint8_t *, size_t), void *env) {
     parse_hci_info_t parse = { cb, env, false };
-    tl_check_msg_ble(&ipcc_mem_ble_evt_queue, &parse);
+    // Either this is being called
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:603]: not sure what this comment means...?

File: ports/stm32/rfcore.c [hunk 6]
```diff
@@ -556,11 +555,15 @@ void rfcore_ble_init(void) {
 
     // Clear any outstanding messages from ipcc_init
     tl_check_msg(&ipcc_mem_sys_queue, IPCC_CH_SYS, NULL);
-    tl_check_msg_ble(&ipcc_mem_ble_evt_queue, NULL);
+    tl_check_msg(&ipcc_mem_ble_evt_queue, IPCC_CH_BLE, NULL);
```
Comment (dpgeorge) [inline ports/stm32/rfcore.c:558]: is this line even needed anymore, now that events are handled immediately in the RX IRQ?

**Comment by dpgeorge** [comment]
Now that RX IRQs are executed immediately, it could be that IPCC IRQ calls `mp_bluetooth_on_hci_irqs` calls `mp_bluetooth_hci_poll_uart` calls `mp_bluetooth_nimble_hci_uart_process` calls into NimBLE, even if the BLE stack (nimble) is disabled.... does this matter, does it need a guard?

File: tests/multi_bluetooth/stress_log_filesystem.py (hunk 4 above)
Comment (jimmo) [reply tests/multi_bluetooth/stress_log_filesystem.py:142]: This file will not be exactly the same across runs.

**Comment by jimmo** [comment]
Rebased on latest updates to #6594. We may need to introduce an RX buffer for rfcore.c as there could be some delay between the IRQ arriving and the scheduler executing and it might be possible for the payload to be replaced before being processed. However with the current changes the tests (including the new one) pass fine, so it's already better than what was there before so that might need to be a future improvement. Also added a fix for sending large payloads where NimBLE sends multiple HCI payloads before receiving acks for the previous ones leading to lost packets.

> Now that RX IRQs are executed immediately, it could be that IPCC IRQ calls `mp_bluetooth_on_hci_irqs` calls `mp_bluetooth_hci_poll_uart` calls `mp_bluetooth_nimble_hci_uart_process` calls into NimBLE, even if the BLE stack (nimble) is disabled.... does this matter, does it need a guard?

No longer required because NimBLE now runs entirely in the scheduler.

File: tests/multi_bluetooth/stress_log_filesystem.py (hunk 4 above)
Comment (dpgeorge) [reply tests/multi_bluetooth/stress_log_filesystem.py:141]: Ok

**Comment by dpgeorge** [comment]
This was rebased and merged in 7e75245d549d01ac4f01ee615c193c00cf3269a7 through 21c293fbcd1f9d5bafac300b5da265b4efd2be52

Thank you!

**Pull request closed without merging**